### PR TITLE
fixes for blob support,ignoring extra tables and perf improvement

### PR DESCRIPTION
**Pull request by aksharauke**
The PR fixes following issues with reverse replication
Reader template:
1. The BLOB type support on DELETE is fixed
2. Tables not in session file are not passed to custom plugin
3. Pipeline sub-optimal fused stages are unfused

Writer template:
1. Skipped file interval tracking is removed since it was bottleneck
2. Timer is moved after setting the next start interval, else files were skipped
3. Fixed stack overflow issue when the reader template is slow in writing files

Both templates:
1. SpannerConfig is used to create the SpannerDao
2. AdminRequest to get the DB dialect is done only once
4. Retries for appropriate Spanner error codes added to prevent exceptions

### CODE REVIEW

**Comment by aksharauke** [comment]
@Dippatel98 @Abacn could you please review these changes?

File: v2/gcs-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/utils/GCSReader.java [hunk 1]
```diff
@@ -204,4 +218,49 @@ private List<TrimmedShardedDataChangeRecord> checkAndReturnIfFileExists() {
           " Cannot determine file creation progress for shard : " + shardId, e);
     }
   }
+
+  private List<TrimmedShardedDataChangeRecord> waitTillFileCreatedAndReturn() {
+    boolean found = false;
+    List<TrimmedShardedDataChangeRecord> changeStreamList = new ArrayList<>();
+    while (!found) {
+      try (InputStream stream =
```
Comment (shreyakhajanchi) [inline v2/gcs-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/utils/GCSReader.java:226]: Existing code from getRecords() can be refactored and reused here

**Review by shreyakhajanchi — APPROVED** [review]
LGTM. Minor refactoring comments

File: v2/spanner-change-streams-to-sharded-file-sink/src/main/java/com/google/cloud/teleport/v2/templates/transforms/AssignShardIdFn.java [hunk 2]
```diff
@@ -497,6 +529,7 @@ private Object getColumnValueFromRow(Column column, Value value) throws Exceptio
         case PG_JSONB:
           return value.getString();
         case BYTES:
+          return value.getBytes();
         case PG_BYTEA:
           return value.getBytesArray();
```
[the remark this answers is not in the document]
Comment (aksharauke) [reply v2/spanner-change-streams-to-sharded-file-sink/src/main/java/com/google/cloud/teleport/v2/templates/transforms/AssignShardIdFn.java:534]: As per this https://source.corp.google.com/piper///depot/google3/third_party/java_src/cloud/teleport/v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/type/Type.java;rcl=601124801;l=65 and https://source.corp.google.com/piper///depot/google3/third_party/java_src/cloud/teleport/v2/spanner-common/src/main/java/com/google/cloud/teleport/v2/spanner/type/Type.java;rcl=601124801;l=78 it is BYTES_ARRAY for PG_BYTEA

File: v2/gcs-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/utils/GCSReader.java (hunk 1 above)
Comment (aksharauke) [reply v2/gcs-to-sourcedb/src/main/java/com/google/cloud/teleport/v2/templates/utils/GCSReader.java:226]: Could not find a good way to do that. But have opened a general bug for code-refactoring during integration testing, will handle there.

File: v2/spanner-change-streams-to-sharded-file-sink/src/main/java/com/google/cloud/teleport/v2/templates/transforms/AssignShardIdFn.java [hunk 3]
```diff
@@ -513,4 +546,27 @@ private Object getColumnValueFromRow(Column column, Value value) throws Exceptio
       throw new Exception("Error getting column value from row: " + e.getMessage());
     }
   }
+
+  private String getShardIdColumnForTableName(String tableName) throws IllegalArgumentException {
```
[the remark this answers is not in the document]
Comment (aksharauke) [reply v2/spanner-change-streams-to-sharded-file-sink/src/main/java/com/google/cloud/teleport/v2/templates/transforms/AssignShardIdFn.java:550]: Have opened a general bug for code-refactoring during integration testing, will handle there.

**Comment by aksharauke** [comment]
@Dippatel98 @Abacn @Polber could someone please help with the LGTM, need to ship this to customer this week. Thanks!

**Pull request merged**